// @cloudcomponents/cdk-dynamodb-seeder/lib/seeds.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonFileSeeds = exports.InlineSeeds = exports.S3Seeds = exports.Seeds = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const fs = require("fs");
const aws_cdk_lib_1 = require("aws-cdk-lib");
class Seeds {
    /**
     * @param bucket The S3 bucket.
     * @param key The object key.
     * @param objectVersion Optional S3 object version.
     * @returns `S3Seeds` associated with the specified S3 object.
     */
    static fromBucket(bucket, key, objectVersion) {
        return new S3Seeds(bucket, key, objectVersion);
    }
    /**
     * @param seeds The actual json code (limited to 4KiB).
     * @returns `InlineSeeds` with inline seeds.
     */
    static fromInline(seeds) {
        return new InlineSeeds(JSON.stringify(seeds));
    }
    /**
     * Loads the seeds from a local disk path and uploads it to s3.
     *
     * @param path Path to json seeds file.
     * @returns `JsonFileSeeds` associated with the specified S3 object.
     */
    static fromJsonFile(path, options) {
        return new JsonFileSeeds(path, options);
    }
}
exports.Seeds = Seeds;
_a = JSII_RTTI_SYMBOL_1;
Seeds[_a] = { fqn: "@cloudcomponents/cdk-dynamodb-seeder.Seeds", version: "2.1.0" };
/**
 * Seeds from an S3 archive.
 */
class S3Seeds extends Seeds {
    constructor(bucket, key, objectVersion) {
        super();
        this.key = key;
        this.objectVersion = objectVersion;
        if (!bucket.bucketName) {
            throw new Error('BucketName is undefined for the provided bucket');
        }
        this.bucketName = bucket.bucketName;
    }
    /**
     * Called when the seeder is initialized to allow this object to bind to the stack.
     */
    bind(_scope) {
        return {
            s3Location: {
                bucketName: this.bucketName,
                objectKey: this.key,
                objectVersion: this.objectVersion,
            },
        };
    }
}
exports.S3Seeds = S3Seeds;
_b = JSII_RTTI_SYMBOL_1;
S3Seeds[_b] = { fqn: "@cloudcomponents/cdk-dynamodb-seeder.S3Seeds", version: "2.1.0" };
/**
 * Seeds from an inline json object (limited to 4KiB).
 */
class InlineSeeds extends Seeds {
    constructor(seeds) {
        super();
        this.seeds = seeds;
        if (seeds.length === 0) {
            throw new Error('Inline seeds cannot be empty');
        }
        if (seeds.length > 4096) {
            throw new Error(`Inline seeds are too large, must be <= 4096 but is ${seeds.length}`);
        }
    }
    /**
     * Called when the seeder is initialized to allow this object to bind to the stack.
     */
    bind(_scope) {
        return {
            inlineSeeds: this.seeds,
        };
    }
}
exports.InlineSeeds = InlineSeeds;
_c = JSII_RTTI_SYMBOL_1;
InlineSeeds[_c] = { fqn: "@cloudcomponents/cdk-dynamodb-seeder.InlineSeeds", version: "2.1.0" };
/**
 * Seeds from a local json file.
 */
class JsonFileSeeds extends Seeds {
    constructor(path, options = {}) {
        super();
        this.path = path;
        this.options = options;
    }
    /**
     * Called when the seeder is initialized to allow this object to bind to the stack.
     */
    bind(scope) {
        // If the same JsonFileSeeds is used multiple times, retain only the first instantiation.
        if (!this.asset) {
            this.asset = new aws_cdk_lib_1.aws_s3_assets.Asset(scope, 'JsonFileSeeds', {
                path: this.path,
                ...this.options,
            });
        }
        try {
            JSON.parse(fs.readFileSync(this.path, 'utf-8'));
        }
        catch (e) {
            throw new Error('Could not convert file to JSON');
        }
        return {
            s3Location: {
                bucketName: this.asset.s3BucketName,
                objectKey: this.asset.s3ObjectKey,
            },
        };
    }
}
exports.JsonFileSeeds = JsonFileSeeds;
_d = JSII_RTTI_SYMBOL_1;
JsonFileSeeds[_d] = { fqn: "@cloudcomponents/cdk-dynamodb-seeder.JsonFileSeeds", version: "2.1.0" };
